### PR TITLE
[SPARK-41159][SQL] Optimize like any and like all expressions

### PR DESCRIPTION
**Pull request by wankunde**
### What changes were proposed in this pull request?

Optimize like any and like all expressions with startWith, endWith, contains, equalTo methods.

### Why are the changes needed?

Now like any and like all expressions will be very slow whether enable or disable LikeSimplification rule.
Refer to `org.apache.spark.sql.execution.benchmark.LikeAnyBenchmark`

```
OpenJDK 64-Bit Server VM 1.8.0_352-b08 on Linux 5.15.0-1022-azure
Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz
[info] Multi like query:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Query with multi like                              1439           1518          78          0.0     1438904.6       1.0X
[info] Query with LikeAny simplification                  1392           1427          30          0.0     1392103.7       1.0X
[info] Query without LikeAny simplification                368            374           5          0.0      368485.2       3.9X
```

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Exists UT

### CODE REVIEW

**Comment by wankunde** [comment]
Hi, @beliefer @cloud-fan @wangyum Could you help to review this PR? Thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 1]
```diff
@@ -248,13 +248,90 @@ case class ILike(
   }
 }
 
+case class MatchMultiHelper(
+    isNotSpecified: Boolean,
+    partialMatch: Boolean,
+    patterns: Seq[UTF8String]) {
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:254]: `LikeSimplification` have the similar optimization. Why need this class ?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/LikeAnyBenchmark.scala [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+ * To run this benchmark:
+ * {{{
+ *   1. without sbt: bin/spark-submit --class <this class>
+ *     --jars <spark core test jar>,<spark catalyst test jar> <spark sql test jar>
+ *   2. build/sbt "sql/Test/runMain <this class>"
+ *   3. generate result: SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "sql/Test/runMain <this class>"
+ *      Results will be written to "benchmarks/LikeAnyBenchmark-results.txt".
+ * }}}
+ */
+object LikeAnyBenchmark extends SqlBasedBenchmark {
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/LikeAnyBenchmark.scala:40]: The benchmark cannot prove the performance improvement. Could you test with or without `MatchMultiHelper` ?

Comment (wankunde) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/LikeAnyBenchmark.scala:40]: Before this PR:
```
[info] Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
[info] Multi like query:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Query with multi like                              1393           1469         119          0.0     1392586.7       1.0X
[info] Query with LikeAny simplification                  1244           1309          97          0.0     1244382.5       1.1X
[info] Query without LikeAny simplification                400            407           8          0.0      399924.3       3.5X

[info] Multi like query:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Query with multi like                              1476           1576         149          0.0     1475710.1       1.0X
[info] Query with LikeAny simplification                  1387           1429          37          0.0     1386669.1       1.1X
[info] Query without LikeAny simplification                430            470          35          0.0      430435.8       3.4X
```

After this PR:
```
[info] Multi like query:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Query with multi like                              1441           1516          78          0.0     1441335.8       1.0X
[info] Query with LikeAny simplification                  1401           1431          44          0.0     1400743.9       1.0X
[info] Query without LikeAny simplification                357            369          10          0.0      357419.8       4.0X

[info] Multi like query:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Query with multi like                              1524           1628         117          0.0     1524119.6       1.0X
[info] Query with LikeAny simplification                  1405           1418          18          0.0     1405258.7       1.1X
[info] Query without LikeAny simplification                362            372          12          0.0      361654.4       4.2X
```

Comment (beliefer) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/LikeAnyBenchmark.scala:40]: It means regression ?

**Comment by wankunde** [comment]
After `LikeSimplification`,  the combination of multiple like expressions with `OR` can be pushdown to parquet reader, while `like any` can not.
So close this PR.

**Pull request closed without merging**